pkg/awsutils/awsutils.go

```diff
@@ -639,38 +639,40 @@ func (cache *EC2InstanceMetadataCache) getIPsAndCIDR(eniMAC string) ([]*ec2.Netw
 }
 
 // getENIDeviceNumber returns ENI ID, device number, error
-func (cache *EC2InstanceMetadataCache) getENIDeviceNumber(eniMAC string) (string, int, error) {
+func (cache *EC2InstanceMetadataCache) getENIDeviceNumber(eniMAC string) (eniID string, deviceNumber int, err error) {
 	// get device-number
 	start := time.Now()
 	device, err := cache.ec2Metadata.GetMetadata(metadataMACPath + eniMAC + metadataDeviceNum)
 	awsAPILatency.WithLabelValues("GetMetadata", fmt.Sprint(err != nil)).Observe(msSince(start))
 	if err != nil {
 		awsAPIErrInc("GetMetadata", err)
 		log.Errorf("Failed to retrieve the device-number of ENI %s, %v", eniMAC, err)
-		return "", 0, errors.Wrapf(err, "failed to retrieve device-number for ENI %s",
-			eniMAC)
+		return "", -1, errors.Wrapf(err, "failed to retrieve device-number for ENI %s", eniMAC)
 	}
 
-	deviceNum, err := strconv.ParseInt(device, 0, 32)
+	deviceNumber, err = strconv.Atoi(device)
 	if err != nil {
-		return "", 0, errors.Wrapf(err, "invalid device %s for ENI %s", device, eniMAC)
+		return "", -1, errors.Wrapf(err, "invalid device %s for ENI %s", device, eniMAC)
 	}
 
 	start = time.Now()
-	eni, err := cache.ec2Metadata.GetMetadata(metadataMACPath + eniMAC + metadataInterface)
+	eniID, err = cache.ec2Metadata.GetMetadata(metadataMACPath + eniMAC + metadataInterface)
 	awsAPILatency.WithLabelValues("GetMetadata", fmt.Sprint(err != nil)).Observe(msSince(start))
 	if err != nil {
 		awsAPIErrInc("GetMetadata", err)
 		log.Errorf("Failed to retrieve the interface-id data from instance metadata service, %v", err)
-		return "", 0, errors.Wrapf(err, "get attached ENIs: failed to retrieve interface-id for ENI %s", eniMAC)
+		return "", -1, errors.Wrapf(err, "get attached ENIs: failed to retrieve interface-id for ENI %s", eniMAC)
 	}
 
-	if cache.primaryENI == eni {
-		log.Debugf("Using device number 0 for primary eni: %s", eni)
-		return eni, 0, nil
+	if cache.primaryENI == eniID {
+		log.Debugf("Using device number 0 for primary ENI: %s", eniID)
+		if deviceNumber != 0 {
+			log.Warnf("Device number of primary ENI is %d! It was expected to be 0", deviceNumber)
+		}
+		return eniID, deviceNumber, nil
 	}
-	// 0 is reserved for primary ENI, the rest of them has to +1 to avoid collision at 0
-	return eni, int(deviceNum + 1), nil
```

Comment:
@SaranBalaji90 Before this PR, `eth1` gets `table 2` because of this code, since it has device number 1. `eth2` gets `table 3` and so on.

Reply:
Are you sure? Because on a c5.18xlarge eth1 got 15 but rest of them got similar number as device index - for eg eth2 got route table number 2 and so on.

Reply:
Was the device index of `eth1` 14 at that time? The device index does not always match the interface name, especially when deleting and attaching a lot of ENIs. I have also seen `eth4` get `table 3` since it had the device number 2.

Reply:
I don't have the node now but I think looking at the code it should be 14. I wonder how is this number populated? Because of pod ENI feature we don't want this number to go beyond 100.

Reply:
Looks like for eth1 for some reason device number is always at the end. Rest of them gets (interface device index - 1) as device number.

Reply:
> Because of pod ENI feature we don't want this number to go beyond 100

I know we're assuming that currently, but do we enforce that in the code anywhere?  It sounds like this warrants a nice loud/obvious error message - if we ever accidentally trigger it.

Fwiw, I think [the relevant code](https://github.com/aws/amazon-vpc-cni-k8s/blob/561190a6f982dbdefea6d233939960366a10397c/pkg/awsutils/awsutils.go#L676-L716) currently always looks for the lowest unoccupied DeviceNumber, so it will continue to reuse the same low numbers (bounded indirectly by max-eni).  In addition, I note the relevant bit of code imposes an [artifical max of 128](https://github.com/aws/amazon-vpc-cni-k8s/blob/561190a6f982dbdefea6d233939960366a10397c/pkg/awsutils/awsutils.go#L709), for no particular reason (the code uses a fixed size array, whereas it could have dynamically resized the array, or used a sparse hash).  We should probably lower `maxENIs` to 100, with a comment explaining why 100 is actually a real limit we have.

Reply:
Good point Gus. For some reason I missed that we were setting the device index. I agree we should change it to 100 and raise error if it goes beyond 100.

Reply:
Good call, set 100 as a hard limit.